cmd/kubeadm/app/phases/copycerts/copycerts_test.go

```diff
@@ -156,3 +160,123 @@ func TestCertOrKeyNameToSecretName(t *testing.T) {
+		t.Fatal("error creating PKI assets")
+	}
+
+	cs := fakeclient.NewSimpleClientset()
+	if err := UploadCerts(cs, initConfiguration, secretKey); err != nil {
+		t.Fatalf("error uploading certs: %v", err)
+	}
+	rawSecretKey, err := hex.DecodeString(secretKey)
+	secretMap, err := cs.CoreV1().Secrets(metav1.NamespaceSystem).Get(kubeadmconstants.KubeadmCertsSecret, metav1.GetOptions{})
+	if err != nil {
+		t.Fatal("could not fetch secret")
+	}
+	for certName, certPath := range certsToTransfer(initConfiguration) {
+		secretCertData, err := cryptoutil.DecryptBytes(secretMap.Data[certOrKeyNameToSecretName(certName)], rawSecretKey)
+		if err != nil {
+			t.Fatal("error decrypting secret data")
+		}
+		diskCertData, err := ioutil.ReadFile(certPath)
+		if err != nil {
+			t.Fatal("error reading certificate in disk")
+		}
+		// Check that the encrypted contents on the secret match the contents on disk, and that all
+		// the expected certificates are in the secret
+		if string(secretCertData) != string(diskCertData) {
+			t.Fatalf("cert %s does not have the expected contents. contents: %q; expected contents: %q", certName, string(secretCertData), string(diskCertData))
+		}
+	}
+}
+
+func TestDownloadCerts(t *testing.T) {
+	secretKey, err := CreateCertificateKey()
+	if err != nil {
+		t.Fatalf("could not create certificate key: %v", err)
+	}
+
+	// Temporary directory where certificates will be generated
+	tmpdir := testutil.SetupTempDir(t)
+	defer os.RemoveAll(tmpdir)
+	initConfiguration := testutil.GetDefaultInternalConfig(t)
+	initConfiguration.ClusterConfiguration.CertificatesDir = tmpdir
+
+	// Temporary directory where certificates will be downloaded to
+	targetTmpdir := testutil.SetupTempDir(t)
+	defer os.RemoveAll(targetTmpdir)
+	initForDownloadConfiguration := testutil.GetDefaultInternalConfig(t)
+	initForDownloadConfiguration.ClusterConfiguration.CertificatesDir = targetTmpdir
+
+	if err := certs.CreatePKIAssets(initConfiguration); err != nil {
+		t.Fatal("error creating PKI assets")
+	}
+
+	kubeadmCertsSecret := createKubeadmCertsSecret(t, initConfiguration, secretKey)
+	cs := fakeclient.NewSimpleClientset(kubeadmCertsSecret)
+	if err := DownloadCerts(cs, initForDownloadConfiguration, secretKey); err != nil {
+		t.Fatalf("error downloading certs: %v", err)
+	}
+
+	for certName, certPath := range certsToTransfer(initForDownloadConfiguration) {
+		diskCertData, err := ioutil.ReadFile(certPath)
+		if err != nil {
+			t.Fatal("error reading certificate in disk")
+		}
+		// Check that the written files are either certificates or keys, and that they have
+		// the expected permissions
+		_, keyErr := keyutil.ParsePublicKeysPEM(diskCertData)
+		_, certErr := certutil.ParseCertsPEM(diskCertData)
+		switch {
+		case keyErr == nil:
+			if stat, err := os.Stat(certPath); err == nil {
+				if stat.Mode() != keyFileMode {
+					t.Fatalf("key %q should have mode %#o, has %#o", certName, keyFileMode, stat.Mode())
```

Comment:
i think inside this loop, it make sense to use t.Errorf(), so that the loop can continue and print all the cert errors?